Require C++17 for building Stepmania.

bphlipot opened the pull request.

teejusb commented:
For some context. They asked if they could upgrade the C++ Standard version in Stamina Nation. C++20 seemed to have some build issues (shown below), so I suggested that C++17 is sufficient for now.

---

I believe there are only 3 compile errors. Two involve `ThemeMetric::operator==()`, specifically comparing `ThemeMetric<float>` and `int` in a couple places. The other is this line:

```
return std::accumulate(v.begin(), v.end(), RString(), [](RString &res, Command const &c) { return res + c.GetOriginalCommandString(); });
```

I'm guessing it has something to do with RString::operator+(), but I got lost/scared looking at the custom string classes. 

---

Assuming the CI builds successfully, I'm okay to merge this.

bphlipot commented:
Welp, it only compiles in the environment I tested in...I guess there's some more work to do.

bphlipot commented:
Is there a way up force it to rerun the CI? Looks like the latest attempt failed because of something server side.

teejusb commented:
I re-ran it but it still seems to be failing.

natano commented:
All the files using `std::random_shuffle` have to `#include <algorithm>`.

natano commented on src/GameSoundManager.cpp:
```diff
@@ -318,7 +318,8 @@ static void DoPlayOnceFromDir( RString sPath )
 			order.push_back(i);
 		}
-		std::random_shuffle(order.begin(), order.end());
+                RandomGen rnd;
```
Indentation issue here (should be tabs)

bphlipot committed:
Require C++17 for building Stepmania.

bphlipot commented:
`std::random_shuffle` was removed in C++17, but for whatever reason the code still builds on Linux using that function. I replaced `std::random_shuffle` with `std::shuffle` with the hope that it solves all the build issues on the other platforms.

The APIs for these two functions are slightly different, so I also had to change some other random number generator code. Specifically, I changed the default random number generator used by the C++ and LUA code, so these changes are a bit more impactful than originally intended.

bphlipot committed:
Merge remote-tracking branch 'master/beta' into cxx17

bphlipot commented:
Closing this PR in favor of the [C++20 PR](https://github.com/itgmania/itgmania/pull/57)
